vpr/src/route/check_route.cpp

```diff
@@ -598,40 +610,66 @@ static void check_node_and_range(RRNodeId inode,
 //This check is slow, so it has been moved out of check_route()
 static void check_all_non_configurable_edges(const Netlist<>& net_list, bool is_flat) {
+    const auto& rr_graph = g_vpr_ctx.device().rr_graph;
+
     vtr::ScopedStartFinishTimer timer("Checking to ensure non-configurable edges are legal");
-    auto non_configurable_rr_sets = identify_non_configurable_rr_sets();
+    const t_non_configurable_rr_sets non_configurable_rr_sets = identify_non_configurable_rr_sets();
+
+    // Specifies which RR set each node is part of.
+    vtr::vector<RRNodeId, int> rrnode_set_ids(rr_graph.num_nodes(), -1);
+
+    int non_configurable_rr_set_id = 0;
+    for (const auto& node_set : non_configurable_rr_sets.node_sets) {
+        for (const RRNodeId node_id : node_set) {
+            VTR_ASSERT_SAFE(rrnode_set_ids[node_id] == -1);
+            rrnode_set_ids[node_id] = non_configurable_rr_set_id;
+        }
+        non_configurable_rr_set_id++;
+    }
 
     for (auto net_id : net_list.nets()) {
         check_non_configurable_edges(net_list,
                                      net_id,
                                      non_configurable_rr_sets,
+                                     rrnode_set_ids,
                                      is_flat);
     }
 }
 
-// Checks that the specified routing is legal with respect to non-configurable edges
-//
-//For routing to be legal if *any* non-configurable edge is used, so must *all*
-//other non-configurable edges in the same set
 static bool check_non_configurable_edges(const Netlist<>& net_list,
                                          ParentNetId net,
                                          const t_non_configurable_rr_sets& non_configurable_rr_sets,
+                                         const vtr::vector<RRNodeId, int>& rrnode_set_id,
                                          bool is_flat) {
     const auto& device_ctx = g_vpr_ctx.device();
-    auto& route_ctx = g_vpr_ctx.mutable_routing();
+    const auto& route_ctx = g_vpr_ctx.routing();
 
     if (!route_ctx.route_trees[net]) // no routing
         return true;
 
-    // Collect all the edges used by this net's routing
+    // Collect all the nodes, edges, and RR sets ids used by this net's routing
     std::set<t_node_edge> routing_edges;
     std::set<RRNodeId> routing_nodes;
-    for (auto& rt_node : route_ctx.route_trees[net].value().all_nodes()) {
+    std::set<int> routing_non_configurable_rr_set_ids;
+    for (const RouteTreeNode& rt_node : route_ctx.route_trees[net].value().all_nodes()) {
         routing_nodes.insert(rt_node.inode);
         if (!rt_node.parent())
             continue;
         t_node_edge edge = {rt_node.parent()->inode, rt_node.inode};
         routing_edges.insert(edge);
+
+        if (rrnode_set_id[rt_node.inode] >= 0) {
+            routing_non_configurable_rr_set_ids.insert(rrnode_set_id[rt_node.inode]);
+        }
+    }
+
+    // Copy used non-configurable RR sets
```

Comment:
I suggest beefing up this comment. You are finding the non-configurable RR sets that are used by this net's routing, and copying their data as you will change it as you check that they are used in a consistent way.